### PR TITLE
Water heating coils do not honor system sizing input for heating capacity #5803

### PR DESCRIPTION
**Pull request by wfbuhl**
## Pull request overview

Water heating coils should size based on the Heating Design Capacity input in the Sizing:System object.
Fixes Issue #5803. When specified, the Heating Design Capacity is used to size the water heating coil UA and water flow rate. Small differences in existing input files with water heating coils expected.

### Work Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
- [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
- [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
  - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
  - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
  - NewFeature: This pull request includes code to add a new feature to EnergyPlus
  - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
  - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog
### Review Checklist

This will not be exhaustively relevant to every PR.
- [x] Code style (parentheses padding, variable names)
- [x] Functional code review (it has to work!)
- [x] If defect, results of running current develop vs this branch should exhibit the fix
- [x] CI status: all green or justified
- [x] Performance: CI Linux results include performance check -- verify this
- [x] Unit Test(s)
- C++ checks:
  - [ ] Argument types
  - [ ] If any virtual classes, ensure virtual destructor included, other things
- IDD changes:
  - [ ] Verify naming conventions and styles, memos and notes and defaults
  - [ ] Open windows IDF Editor with modified IDD to check for errors
  - [ ] If transition, add rules to spreadsheet
  - [ ] If transition, add transition source
  - [ ] If transition, update idfs
- [ ] If new idf included, locally check the err file and other outputs
- [ ] Documentation changes in place
- [ ] Changed docs build successfully
- [ ] ExpandObjects changes?
- [ ] If output changes, including tabular output structure, add to output rules file for interfaces

### CODE REVIEW

**Comment by nrel-bot-3** [comment]
@wfbuhl @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@wfbuhl @lgentile it has been 14 days since this pull request was last updated.

**Comment by rraustad** [comment]
Defect file: HelpDesk3-SizingOnly

  Sizing:System,
    OAF-BM1.01,              !- AirLoop Name
    HeatingDesignCapacity,   !- Heating Design Capacity Method
    43500,                   !- Heating Design Capacity {W}

Old result:
Water Heating Coil Capacity Information,Coil:Heating:Water,OAF-BM1.01 HEATING COIL,48316.85
New result:
Water Heating Coil Capacity Information,Coil:Heating:Water,OAF-BM1.01 HEATING COIL,43587.89

  Sizing:System,
    AHU.WP,                  !- AirLoop Name
    HeatingDesignCapacity,   !- Heating Design Capacity Method
    46000,                   !- Heating Design Capacity {W}

Old result:
Water Heating Coil Capacity Information,Coil:Heating:Water,AHU.WP HEATING COIL,65948.70
New result:
Water Heating Coil Capacity Information,Coil:Heating:Water,AHU.WP HEATING COIL,46092.94

  Sizing:System,
    AHU.WI,                  !- AirLoop Name
    HeatingDesignCapacity,   !- Heating Design Capacity Method
    183000,                  !- Heating Design Capacity {W}

Old result:
Water Heating Coil Capacity Information,Coil:Heating:Water,AHU.WI HEATING COIL,187898.87
New result:
Water Heating Coil Capacity Information,Coil:Heating:Water,AHU.WI HEATING COIL,183369.74

There are also changes to eio verbage related to "design size" pre and post change:
 Component Sizing Information, Coil:Heating:Water, AHU.WI HEATING COIL, Design Size Design Coil Load [W], 183000.00000
 Component Sizing Information, Coil:Heating:Water, AHU.WI HEATING COIL, Design Size Maximum Water Flow Rate [m3/s], 3.97564E-003

versus

 Component Sizing Information, Coil:Heating:Water, AHU.WI HEATING COIL, Design Size , 183000.00000
 Component Sizing Information, Coil:Heating:Water, AHU.WI HEATING COIL, Maximum Water Flow Rate [m3/s], 4.00674E-003

File: src/EnergyPlus/WaterCoils.cc
Comment (rraustad) [inline src/EnergyPlus/WaterCoils.cc]: @wfbuhl , what if the user enters a value for MaxWaterVolFlowRate? Are we just going to overwrite that value?

**Comment by rraustad** [comment]
I made a slight change to the sizing procedure to allow a user to enter a water flow rate when specifying HeatingDesignCapacity as the sizing method. I need @wfbuhl to comment on whether a user input for water flow rate is allowed for this sizing method. I used 0.0007 m3/s as the hard input yielding this result. U-factor before I used 0.0007 was 869 so this looks like it still works as expected (lower flow, raise U). 

```
Component Sizing Information, Coil:Heating:Water, OAF-BM1.01 HEATING COIL, Design Size Rated Capacity [W], 43500.00000
Component Sizing Information, Coil:Heating:Water, OAF-BM1.01 HEATING COIL, Design Size Maximum Water Flow Rate [m3/s], 9.52421E-004
Component Sizing Information, Coil:Heating:Water, OAF-BM1.01 HEATING COIL, User-Specified Maximum Water Flow Rate [m3/s], 7.00000E-004
Component Sizing Information, Coil:Heating:Water, OAF-BM1.01 HEATING COIL, Design Size U-Factor Times Area Value [W/K], 920.33534
Water Heating Coil Capacity Information,Coil:Heating:Water,OAF-BM1.01 HEATING COIL,43621.59
```

**Comment by rraustad** [comment]
These changes moves coil sizing much closer to user specified inputs. I'm leaning toward merging since it's better than before. Waiting for comments.

File: src/EnergyPlus/WaterCoils.cc
Comment (rraustad) [inline src/EnergyPlus/WaterCoils.cc]: @wfbuhl these were necessary before because an input file using Performance Input Method = UFactorTimesAreaAndDesignWaterFlowRate where U-Factor and Max Water Flow were not autosized did not require a Sizing:Plant object. Now some example files are showing a warning where a Sizing:Plant object could not be found. Whether these are used or not the Sizing routine needs to account for when a sizing run is required and when it's not.

Comment (rraustad) [reply src/EnergyPlus/WaterCoils.cc]: 5ZoneDDCycOnAny is one example.

**Comment by rraustad** [comment]
@Myoldmopar unhandled failure above? x86_64-Linux-Ubuntu-14.04-gcc-4.8

**Comment by nrel-bot-3** [comment]
@wfbuhl @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@wfbuhl @lgentile it has been 15 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@wfbuhl @lgentile it has been 15 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@wfbuhl @lgentile it has been 15 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@wfbuhl @lgentile it has been 14 days since this pull request was last updated.

**Comment by mjwitte** [comment]
@rraustad What is the status of this?  I has some conflicts that need to be fixed, but is the basic change ready to go or does it need more work?  Your last review comments in November said you were leaning towards merging.

**Comment by rraustad** [comment]
Yes I was ready to merge, then looked at the results and some files are no longer sizing. I was going to circle back to this as soon as I could find time.

**Comment by rraustad** [comment]
Starting on this now. Conflicts resolved. Looking into what else needs to be done to correct some example file sizing issues.

**Comment by rraustad** [comment]
This is the error now with some example files. This file, CentralChillerHeaterSystem_Cooling_Heating runs in V8.6 and fails now.

    ** Severe  ** For autosizing of Coil:Heating:Water REHEAT COIL ZONE 1, a zone sizing run must be done.
    **   ~~~   ** No "Sizing:Zone" objects were entered.
    **   ~~~   ** The "SimulationControl" object did not have the field "Do Zone Sizing Calculation" set to Yes.

This coil should not need a sizing run since UA and water flow are specified. 

  Coil:Heating:Water,
    Reheat Coil Zone 1,      !- Name
    400.0,                   !- U-Factor Times Area Value {W/K}
    0.0013,                  !- Maximum Water Flow Rate {m3/s}
    UFactorTimesAreaAndDesignWaterFlowRate,  !- Performance Input Method
    autosize,                !- Rated Capacity {W}

**Comment by rraustad** [comment]
@Myoldmopar @mjwitte many files are failing due to a change in input field name. The old sizing method used a label for the call to ReportSizing, the new method uses the actual field name. When the python script tests for differences in the eio file, I guess it tries to match strings within the file. Since I can't see the differences for these files I am running locally (which we should do anyway but it takes extra time).

Here is what the CI shows (5ZoneAirCooled - see last line):

failed regression.5ZoneAirCooled 3.57254 
 Device id: Win64-Windows-7-VisualStudio-14
Comparing `C:/CI/EnergyPlus-6cd1d04aa7-Win64-VS14/build/testfiles/5ZoneAirCooled` with `C:/ci/EnergyPlus-f7b4f06d6c-Win64-VS14/build/testfiles/5ZoneAirCooled`
Processing (Diffs) : 5ZoneAirCooled
Traceback (most recent call last):
 File "C:/CI/EnergyPlus-f7b4f06d6c-Win64-VS14/regressions/Testing/RegressionSuite/Scripts/RunComparison.py", line 78, in 
 runner.process_diffs_for_one_case(entry, base_dir, mod_dir, mod_dir)
 File "C:\CI\EnergyPlus-f7b4f06d6c-Win64-VS14\regressions\Testing\RegressionSuite\Scripts\runtests.py", line 445, in process_diffs_for_one_case
 path_to_tablediff_log)))
 File "C:\CI\EnergyPlus-f7b4f06d6c-Win64-VS14\regressions\Testing\RegressionSuite\Scripts\TableDiff.py", line 434, in table_diff
 hdict2soup(abs_diff_soup, uheading1, count_of_tables, abs_diff_dict, h_thresh_dict, horder1)
 File "C:\CI\EnergyPlus-f7b4f06d6c-Win64-VS14\regressions\Testing\RegressionSuite\Scripts\TableDiff.py", line 208, in hdict2soup
 (diff, which) = hdict[h][i]
KeyError: u'Design Size Design Coil Load [W]'
 
Here is the WinDiff of the before and after:

fancoilfailures

**Comment by rraustad** [comment]
Updated defect file (5803-HelpDesk3-SizingOnly40k) results:

    Sizing:System,
      OAF-BM1.01,              !- AirLoop Name
      HeatingDesignCapacity,   !- Heating Design Capacity Method
      43500,                   !- Heating Design Capacity {W}

    Component Sizing Information, Coil:Heating:Water, OAF-BM1.01 HEATING COIL, Design Size Rated Capacity [W], 43500.00000
    Component Sizing Information, Coil:Heating:Water, OAF-BM1.01 HEATING COIL, Design Size Maximum Water Flow Rate [m3/s], 9.52421E-004
    Component Sizing Information, Coil:Heating:Water, OAF-BM1.01 HEATING COIL, Design Size U-Factor Times Area Value [W/K], 869.37254
    Water Heating Coil Capacity Information,Coil:Heating:Water,OAF-BM1.01 HEATING COIL,43587.89

**Comment by rraustad** [comment]
There is a new warning in VentilatedSlab_SeriesSlabs as:

     ** Warning ** GetDensityGlycol: Temperature is out of range (too high) for fluid [WATER] density **
     **   ~~~   ** ..Called From:PlantPumps:CalcPumps: ,Temperature=[115.55], supplied data range=[0.00,100.00]
     **   ~~~   **  Environment=CHICAGO_IL_USA ANNUAL HEATING 99% DESIGN CONDITIONS DB, at Simulation time=01/21 00:00 - 00:10
     ************* Simulation Error Summary *************
     *************
     ************* ===== Recurring Error Summary =====
     ************* The following recurring error messages occurred.
     *************
     *************  ** Warning ** Plant loop exceeding upper temperature limit, PlantLoop="HOT WATER LOOP"
     *************  **   ~~~   **   This error occurred 2648 total times;
     *************  **   ~~~   **   during Warmup 1818 times;
     *************  **   ~~~   **   during Sizing 0 times.
     *************  **   ~~~   **   Max=117.898773
     *************
     *************  ** Warning ** GetDensityGlycol: Temperature out of range (too high) for fluid [WATER] density **
     *************  **   ~~~   **   This error occurred 9792 total times;
     *************  **   ~~~   **   during Warmup 0 times;
     *************  **   ~~~   **   during Sizing 0 times.
     *************  **   ~~~   **   Max=117.924034 {C}  Min=115.442671 {C}

I don't think this is directly related to this change to water heating coil sizing although apparently it is in some way. Also, the heating coil size now is smaller, not larger ( which I would have thought would lead to higher water temps).

  Coil:Heating:Water,
    Zone1VentSlabHeatingCoil,!- Name
    FanAndCoilAvailSched,    !- Availability Schedule Name
    400.,                    !- U-Factor Times Area Value {W/K}
    0.8,                     !- Maximum Water Flow Rate {m3/s}
    Zone1VentSlabHWInletNode,!- Water Inlet Node Name
    Zone1VentSlabHWOutletNode,  !- Water Outlet Node Name
    Zone1VentSlabCCOutletNode,  !- Air Inlet Node Name
    Zone1VentslabSlabInNode, !- Air Outlet Node Name
    NominalCapacity,         !- Performance Input Method
    40000,                   !- Rated Capacity {W}

Old:
Water Heating Coil Capacity Information,Coil:Heating:Water,ZONE1VENTSLABHEATINGCOIL,39986.42
New:
Water Heating Coil Capacity Information,Coil:Heating:Water,ZONE1VENTSLABHEATINGCOIL,21639.95

So the rated capacity is now much lower. I suspect that this result should be closer to the user input of 40kW as it was before?

The user has set the 3 model parameters and specified the air flow in the ventilated slab.

    ZoneHVAC:VentilatedSlab,
      Zone1VentSlab,           !- Name
      0.84,                    !- Maximum Air Flow Rate {m3/s}
      Coil:Heating:Water,      !- Heating Coil Object Type
      Zone1VentSlabHeatingCoil,!- Heating Coil Name

So how does the user know that the Rated Capacity is 40kW given air and water flow rate and UA and rated air and water inlet conditions ??? The Rated Capacity (DesTotWaterCoilLoad) isn't even used unless the UA or water flow rate is autosized so it means nothing to the model with these inputs.

**Comment by rraustad** [comment]
Everything I've looked at as far as diff's is the same as what is shown above in the WinDiff figure.
 - the Design Size Rated Capacity has not changed (except the string)
 - the Design Size U-Factor Times Area Value is slightly larger
 - and the resulting Water Heating Coil Capacity is slightly larger

Nothing I can find to hold up this merge.

**Comment by rraustad** [comment]
So will I be merging this or does someone else need to review my changes/conclusions ?

**Comment by rraustad** [comment]
Looked at the VentilatedSlab_SeriesSlab results again. This is weird. There is only a demand side HW coil and supply side district heating and pump on the HW loop. The pump has 0 delta T and HW coil has a negative delta T. If the coil is cooling the water and pump is not adding heat then where are these high temperatures coming from? Must be from control of district heating?

ventslabdt

You have to wonder sometimes why developers choose their inputs.

     Pump:VariableSpeed,
     HW Circ Pump,            !- Name
     HW Supply Inlet Node,    !- Inlet Node Name
     HW Pump Outlet Node,     !- Outlet Node Name
     0.002,                   !- Design Maximum Flow Rate {m3/s}
     300000,                  !- Design Pump Head {Pa}
     900,                     !- Design Power Consumption {W}
     0.87,                    !- Motor Efficiency
     0.0,                     !- Fraction of Motor Inefficiencies to Fluid Stream
     0,                       !- Coefficient 1 of the Part Load Performance Curve
     1,                       !- Coefficient 2 of the Part Load Performance Curve
     0,                       !- Coefficient 3 of the Part Load Performance Curve
     0,                       !- Coefficient 4 of the Part Load Performance Curve
     0,                       !- Design Minimum Flow Rate {m3/s}
     INTERMITTENT;            !- Pump Control Type

**Comment by mjwitte** [comment]
@rraustad Is the ventilated slab concern holding this up?

**Comment by rraustad** [comment]
@Myoldmopar @mjwitte @lgu1234 I have revised Fred's code changes to allow the example files to size without a Sizing:Zone object (only for those files that didn't need one) and now think this is ready to merge. I didn't know if I should do that since I made the final changes to the code. The results from the defect file look correct now (see above 7 days ago just below the pic of the WinDiff results). The ventilated slab results don't concern me that much except that the cause of high water temps needs to be identified at some point. It's a strange example file anyway.

**Comment by mjwitte** [comment]
Yeah, someone else should review and merge. Will put it on the list.

**Comment by nrel-bot-3** [comment]
@rraustad @wfbuhl @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@rraustad @wfbuhl @lgentile it has been 14 days since this pull request was last updated.

**Comment by Myoldmopar** [comment]
I'm back online, merging latest develop in here while I review.

**Comment by Myoldmopar** [comment]
Yeah, after merging in develop, the new CI report is nicely showing the unit test failures that need to be fixed up.  After you are satisfied those are fixed up I'll review the regressions and we can work to get this merged in.

**Comment by rraustad** [comment]
All instances of CalcFinalZoneSizing were removed from the zone equipment sizing routines. This branch has new unit tests that still used those variables. CI should be clean now.

**Comment by rraustad** [comment]
CI does not like it if a value in the eio has changed AND the field name also changed. Here, the field name Design Coil Load was changed to the Coil:Heating:Water input field for Rated Capacity.

**Comment by Myoldmopar** [comment]
Yeah, I mean CI has a lot of diffs, but it seems like they are justifiable.  Code review next...

File: src/EnergyPlus/WaterCoils.cc
Comment (Myoldmopar) [inline src/EnergyPlus/WaterCoils.cc:2103]: Yikes.  Tricky.

File: src/EnergyPlus/WaterCoils.cc
Comment (Myoldmopar) [inline src/EnergyPlus/WaterCoils.cc]: You fixed like every bad padding but then this little one snuck in.

File: src/EnergyPlus/WaterCoils.cc
Comment (Myoldmopar) [inline src/EnergyPlus/WaterCoils.cc:2043]: It's not showing in this diff, so can you confirm that you are protecting for `(DesOutletAirTemp==DesInletAirTemp)` somewhere in before these calculations?

File: src/EnergyPlus/WaterCoils.cc
Comment (Myoldmopar) [inline src/EnergyPlus/WaterCoils.cc]: If you're fixing anything else here before this merges, could you pull the `else` blocks up?  This style is very lengthy, and reduces the amount of logic that can be seen on the screen at once.

File: src/EnergyPlus/WaterCoils.cc
Comment (Myoldmopar) [inline src/EnergyPlus/WaterCoils.cc:2087]: If you don't need this, why do it?

**Review by Myoldmopar — CHANGES_REQUESTED** [review]
There were minor code changes, mostly style, requested.  More importantly are just a few talking points to get through.  Overall this is likely ready to go in as a solid fix.

**Comment by Myoldmopar** [comment]
I had left some minor comments here, are you able to address those and I'll work to merge this in?

**Comment by rraustad** [comment]
Git is back up. I'll look at these comments.

File: src/EnergyPlus/WaterCoils.cc
Comment (rraustad) [reply src/EnergyPlus/WaterCoils.cc:2043]: GetInput tests this for the heating coil and these fields are not autosizable. I do see other instances of a possible divide by 0 for the water coil but those have always been there.

    if ( WaterCoil( CoilNum ).DesInletAirTemp >= WaterCoil( CoilNum ).DesOutletAirTemp ) {
        ShowSevereError( "For " + CurrentModuleObject + ", " + AlphArray( 1 ) );
        ShowContinueError( "  the " + cNumericFields( 5 ) + " must be less than the " + cNumericFields( 7 ) + '.' );
        ErrorsFound = true;
    }

File: src/EnergyPlus/WaterCoils.cc
Comment (rraustad) [reply src/EnergyPlus/WaterCoils.cc:2087]: These variables don't need to be reported, but they do need to be calculated.

**Comment by rraustad** [comment]
I fixed these, committed and pushed (twice). I don't see them here yet so I guess Git is still having issues.

**Comment by Myoldmopar** [comment]
Yeah, that's weird, they still aren't showing up.  When your back on, try pushing again and see if you can get any response out of your Git client so I can help debug any issues.

**Comment by rraustad** [comment]
It says I'm up to date on the Push. Same as yesterday when I tried the second time.

push

**Comment by Myoldmopar** [comment]
And it doesn't show any modified files that haven't been committed yet?  Check the Git log and see if it shows your commits locally.

**Comment by rraustad** [comment]
Hopefully, my previous changes that didn't make it to Git and this merge bring this branch up to where I was before.

**Comment by rraustad** [comment]
@Myoldmopar I think this is ready now.

**Pull request merged**
